base/tuple.jl

```diff
@@ -615,3 +615,32 @@ empty(@nospecialize x::Tuple) = ()
 
 foreach(f, itr::Tuple) = foldl((_, x) -> (f(x); nothing), itr, init=nothing)
 foreach(f, itrs::Tuple...) = foldl((_, xs) -> (f(xs...); nothing), zip(itrs...), init=nothing)
+
+"""
+    circshift(x::Tuple, shift::Int)
+
+Circularly shift, i.e. rotate, the data in a tuple. The second argument specifies
+the shift amount.
+
+# Examples
+```jldoctest
+julia> x = (1, 2, 3, 4, 5)
+(1, 2, 3, 4, 5)
+
+julia> circshift(x, 4)
+(2, 3, 4, 5, 1)
+
+julia> circshift(x, 7)
+(4, 5, 1, 2, 3)
+
+julia> circshift(x, -1)
+(2, 3, 4, 5, 1)
+
+julia> y = ((1, 2), (3, 4), (5, 6))
+((1, 2), (3, 4), (5, 6))
+
+julia> circshift(y, 1)
+((5, 6), (1, 2), (3, 4))
+```
+"""
+circshift(x::Tuple, shift::Int) = ntuple(j -> x[mod1(j-shift, length(x))], Val(length(x)))
```

Comment:
```suggestion
circshift(x::Tuple, shift::Integer) = ntuple(j -> x[mod1(j-shift, length(x))], Val(length(x)))
```
It would be nice to allow any type of integer as the shift. Also, would it produce too much specialization to type the argument as `x::NTuple{N}` and reference `N` in place of `length(x)`?

Reply:
isn't `length` known compile time for `Tuple` ?

Reply:
Yep. Using `NTuple` as I noted would be more for simplicity, but it would force specialization.

Reply:
oh I see. yes looks like `@code_warntype` is happy with that

Reply:
Thank you! I'll checkout and compare `NTuple{N}` based syntax with the current solution.